docs/specs/src/hwp_4_formal.md

```diff
@@ -0,0 +1,534 @@
+
+As such, Holochain doesn't rely on nodes to volunteer to hold specific entries, but rather to volunteer aggregate capacity (e.g., holding 100MB of data rather than arbitrarily chosen entries). Authoring nodes are responsible for publishing entries from their local DHT instance to other nodes on the network (**authorities**) who will become responsible for serving that data.
+
+Like most DHT architectures, Holochain uses a "nearness" algorithm to compute the "distance" distance between the 256-bit Blake2b basis hash of the data or metadata to be stored and the 256-bit Ed25519 public key (network address) of nodes. Basically, it is the responsibility of the nodes *nearest* a basis hash to store data and metadata for it.
+
+Holochain's validating, graphing, gossiping DHT implementation is called **rrDHT**.
+
+rrDHT is designed with a few performance requirements/characteristics in mind.
+
+1. It must have a compact and computationally simple representational model for identifying which nodes are responsible for which content, and which nodes actually hold which content. (A **"world model"** of what is where.)
+2. It must have **lookup speeds** at least as fast as Kademlia's binary trees ( $\mathcal{O}(n \log n)$ ). Current testing shows an average of 3 hops/queries to reach an authority with the data.
+3. It must be **adjustable** to be both resilient and performant across many DHT compositional make-ups (reliability of nodes, different network topologies, high/low usage volumes, etc.)
+
+**World Model:** The network location space is a circle comprising of the range of unsigned 32-bit numbers, in which the location $0$ is adjacent to the location ${2^{32}}-1$.
+
+$$
+\begin{aligned}
+L &: \{l_0,\dots,l_{2^{32}-1}\} \\
+l_{2^{32}-1} &< l_0
+\end{aligned}
+$$
+
+The larger 256-bit address space of the DHT, consisting of 256-bit "basis hashes" $B$ (Blake2b-256 hashes of addressable content $C$, and Ed25519 public keys of agents $K$), is mapped to the smaller network location space via a function $\mathsf{map\_to\_loc}(b)$, which is the XOR of 8 × 32-bit segments of the hashes. The original address is still kept with the data or metadata, so collisions in the smaller space are not a concern.
+
+$$
+\begin{aligned}
+E &: \{e_0,\dots,e_{2^{256}-1}\} \\
+K &: \{k_0,\dots,k_{2^{256}-1}\} \\
+C &: E \cup K \\
+\mathsf{hash} &: C \rightarrow B \\
+\mathsf{map\_to\_loc} &: B \rightarrow L \\
+|L| &\approx |B| \cdot 2^8
```

Comment:
This is confusing and/or incorrect. B (basis hash) is much larger than L (location), so how could multiplying B by something approximate L? And what is the meaning of the absolute value notation here?

Reply:
I got the arithmetic backwards, is all :) As for the squishy equality $\approx$, I was thinking because there would be an arbitrary number of collisions when mapping the larger set to the smaller, _if_ you were dealing with actual sparse addresses on the DHT. Better, maybe, to talk about the address space as a fully allocated thing.